### PR TITLE
add cli tool to query version of package in build

### LINKED ISSUE #278: cli command to get the version of a specific package in komodo release
We should provide an utility to extract a package version used in a specific komodo version. This should be something similar to `pip show <library>` while only extracting the readable version number. The goal is to use it in gh actions workflows to get the proper package version when cloning it.
Eg.
```sh
komodo --show-package ert --release 2022.11.00-py38
>> 4.1.0
```

### PR DESCRIPTION
**Pull request by kwinkunks**
The commit contains a new CLI tool, komodo-show-version (named after pip show, which shows more than just version). It detects the active environment and looks up the version (or git commit hash for non-pip packages) from the 'manifest', a text file komodo creates at build time. Works for komodo and komodoenv environments.

The commit also contains tests.

fixes #278

### CODE REVIEW

File: komodo/show_version.py
Comment (berland) [inline komodo/show_version.py:28]: add `encoding="utf-8"`

File: komodo/show_version.py
Comment (berland) [inline komodo/show_version.py:41]: We don't need to specify the type inside the docstrings, it is (or can be) inferred from the function signature.

File: komodo/show_version.py
Comment (berland) [inline komodo/show_version.py:47]: (`maxsplit=1` is an alternative to `*_`)

File: tests/test_show_version.py
Comment (berland) [inline tests/test_show_version.py:36]: I suppose this test covers the case in https://github.com/equinor/komodo-releases/pull/3043/files#diff-15f0a4593f5be26df91be4cafda7792e49d2c795c2b82ef1e72b58bd8abf89b1R30-R31

File: komodo/show_version.py
Comment (berland) [inline komodo/show_version.py:106]: Add the content of the `release` string to the error message?

File: komodo/show_version.py
Comment (berland) [inline komodo/show_version.py:101]: (Pylint will complain about these one-letter variable names)

File: tests/test_show_version.py
Comment (berland) [inline tests/test_show_version.py:95]: shold -> should

File: komodo/show_version.py
Comment (berland) [inline komodo/show_version.py:19]: We should not use bare exceptions. You may use a dummy exception `KomodoShowVersionException` that just inherits from `Exception` if needed.

**Comment by berland** [comment]
Is this supposed to work or should it fail like this?
```
 [havb@be-lx895926:/work/projects/komodo] cli-get-version$ komodo-show-version  --path /prog/res/komodo/2022.10.03-py38-rhel7/2022.10.03-py38-rhel7  ecl2df
Traceback (most recent call last):
  File "/work/venv/new/bin/komodo-show-version", line 33, in <module>
    sys.exit(load_entry_point('Komodo', 'console_scripts', 'komodo-show-version')())
  File "/work/projects/komodo/komodo/show_version.py", line 129, in main
    return get_version(args.package, args.path)
  File "/work/projects/komodo/komodo/show_version.py", line 101, in get_version
    with open(path / release_file, "rt") as f:
FileNotFoundError: [Errno 2] No such file or directory: '/prog/res/komodo/2022.10.03-py38-rhel7/2022.10'
```

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:41]: Fair point, esp since two of them were wrong :grimacing:

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:47]: Not sure, still requires at least the comma (i.e. `path, = str.split(maxsplit=1)`) because the split always returns at least two items. Will keep current unless you really don't like that syntax.

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:84]: Changed. I'm never sure what to call file paths. I want to be explicit that it wants the **full path** to the 'manifest' file (and name I made up, by the way, but 'release file' is something else).

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:106]: Good one.

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:19]: Changed to `RuntimeError`, what do you think? Definitely an improvement, not sure what's best.

File: tests/test_show_version.py
Comment (kwinkunks) [reply tests/test_show_version.py:95]: fxed

File: tests/test_show_version.py
Comment (kwinkunks) [reply tests/test_show_version.py:36]: :thinking: I'll ask you more about this when I see you.

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:129]: Good point, I'll think about how to avoid that.

**Comment by kwinkunks** [comment]
Will review this one too, thank you Håvard!

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:129]: I think the main situation when we know what's wrong is when there's no $KOMODO_RELEASE. If that was found and the program continues, everything else "should work" and is an exception. So I'll catch that scenario (`KeyError` for the env lookup) and call `sys.exit` in that case.

(`argparse` already deals with mangled/missing args)

Sound reasonable, or did I get the wrong end of the stick @berland ?

File: komodo/show_version.py
Comment (berland) [reply komodo/show_version.py:47]: ah, then keep it :)

File: komodo/show_version.py
Comment (berland) [reply komodo/show_version.py:129]: Sounds good!

File: tests/test_show_version.py
Comment (berland) [reply tests/test_show_version.py:36]: I was specifically looking for a test that covers the case when the env variable `$KOMODO_RELEASE` contains the full path and not just a string, since I used that as a workaround in the link above. And it looks like you have covered it 👍🏻

**Comment by kwinkunks** [comment]
Of course, I wasn't testing passing in the `path` argument, so I earned this last bug with my negligence. Fixed. (And testing for it now :)

File: tests/test_show_version.py
Comment (kwinkunks) [reply tests/test_show_version.py:36]: I think it does.

**Comment by kwinkunks** [comment]
Decided to defer the refactoring to `click` (or maybe `rich-cli`?) and PR this again as-is. I think it still needs approval from someone. Cc @berland @xjules

**Comment by berland** [comment]
This one I would think should work:
```
$  komodo-show-version --path examples/releases/ecl.yml libecl
<snip>
AttributeError: 'str' object has no attribute 'get'
```

**Comment by berland** [comment]
It probably wouldn't hurt to use what's in  `examples` in tests, also to ensure the examples does not get too outdated..

**Comment by kwinkunks** [comment]
`komodo-show-version --path examples/releases/ecl.yml libecl` should actually not work currently, because that's a release 'specification' file, whereas the so-called 'manifest' files I'm expecting are created by `komodo` when it builds the release. It _should_ match the specification, of course... but I think the idea is that it's a less canonical source of the version information.

But it raises the question what this tool should do if someone gives it the specification file. Options are:

- Error (more informatively than I am currently erroring) <- I think this is my preference, but could be convinced otherwise
- Return the version but with a warning because we don't quite trust this source
- Just return the version because we trust this source as much as the manifest

This is also the reason I didn't use `examples` in testing --- there's no example of a manifest file currently, so the test would have to build a komodo release first. Do you think that would be a better approach?

**Comment by berland** [comment]
> But it raises the question what this tool should do if someone gives it the specification file. Options are:
> 
> * Error (more informatively than I am currently erroring) <- I think this is my preference, but could be convinced otherwise

I would go for this, maybe just state that the yaml file has a wrong format, optinally saying the format should be like this and that (then the future developer might deduce which of these files it is meant to work on)

File: tests/test_show_version.py
Comment (xjules) [inline tests/test_show_version.py:185]: sense -> retrieve?

Comment (kwinkunks) [reply tests/test_show_version.py:185]: Agreed :heavy_check_mark:

**Comment by kwinkunks** [comment]
Moved fixtures to `conftest.py` after review from @xjules, see #303.

**Comment by kwinkunks** [comment]
Cannot see how to get `pytest` to see the 'outer' exception in `komodo.show_version.get_version()`. It only wants to see the 'inner' version. This doesn't affect the behaviour of the tool, but it does mean `test_show_version.test_get_version_fails_with_release_file()` is only checking the 'inner' exception.

File: komodo/show_version.py
Comment (hnformentin) [inline komodo/show_version.py:130]: Why does `manifast` need to be in ''?

File: tests/conftest.py
Comment (hnformentin) [inline tests/conftest.py:10]: Why vanilla with '' here and below? Same with 'fake'...

File: tests/test_show_version.py
Comment (hnformentin) [inline tests/test_show_version.py:1]: What about `from pathlib import Path`?

File: tests/test_show_version.py
Comment (hnformentin) [inline tests/test_show_version.py:38]: In this file also there are some uses of '' that seems to be unnecessary.

File: tests/test_show_version.py
Comment (hnformentin) [inline tests/test_show_version.py:31]: `excinfo` is very difficult to pronounce. What do you think about refactoring this name?

**Review by hnformentin — COMMENTED** [review]
Proactive use of docstrings, which is something we are looking for to improve in the team. We had a discussion some time ago where we stressed (1) the use of docstrings to explain why instead of what, and (2) the right granularity, which also implies not to write trivial documentation. More can be found [here](https://github.com/equinor/scout/blob/cf8de3f83b88e262ffcb3162d20bac8c4269392e/testing-and-documentation-guide.md#documentation), if you have time to consider for this PR.

File: komodo/show_version.py
Comment (berland) [reply komodo/show_version.py:4]: I tend to prefer not to import individual functions, as it makes it more readable when that function is in use further down in the code (then I don't have to look at the top to see where that function comes from). But that might come at a speed penalty?

File: komodo/show_version.py
Comment (berland) [reply komodo/show_version.py:35]: I see sometimes this is solved by use of `textwrap.dedent`. Random example: https://github.com/equinor/ert/blob/898ce11efb4d749aea8aba18ab02386486eb201e/tests/unit_tests/c_wrappers/res/enkf/test_run_path_creation.py#L22

File: komodo/show_version.py
Comment (hnformentin) [reply komodo/show_version.py:4]: That is indeed more readable to have the full path. 
I agree that the downside can be speed. Do you think this would make any difference in terms of security? 

There is a way to import one individual function and keep readability (from python documentation):
"""
Users of the package can import individual modules from the package, for example:

import sound.effects.echo
This loads the submodule sound.effects.echo. It must be referenced with its full name.

sound.effects.echo.echofilter(input, output, delay=0.7, atten=4)
"""

How does it look for you?

Comment (kwinkunks) [reply komodo/show_version.py:4]: I'm never sure which way to go with this, but in general I think I prefer more namespace. I don't think it's a speed thing: the entire module is evaluated either way.

It's true you can import submodules with that other pattern, so `import os.path` would be okay, but there's no other way to import `argparse.ArgumentParser` for example. I'm not using any submodules so I'm not sure there's much to do here.

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:57]: Oh right, I meant the environment variable; edited to make it clearer.

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:122]: I did have it like that, but now that I'm parsing the file using `argparse`, this function does not always know the path.

Anyway, I've added the path, if available. But it feels a bit polymorphic...

File: tests/test_show_version.py
Comment (kwinkunks) [reply tests/test_show_version.py:1]: :+1:

File: tests/test_show_version.py
Comment (kwinkunks) [reply tests/test_show_version.py:38]: I guess this is me continuing to struggle with emphasizing that there are two different scenarios. Maybe I should move the `komodoenv` tests to another file, that might help.

Comment (kwinkunks) [reply tests/test_show_version.py:38]: (There are 3 scenarios though, because passing in a manifest file is another option. And I don't think we want 3 test files.)

File: tests/test_show_version.py
Comment (kwinkunks) [reply tests/test_show_version.py:31]: It's the name they use in the `pytest` docs, but I'm happy to change it. `exception_info`?

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:17]: :+1:

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:20]: Changed to "tells us".

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:46]: Done, added to docstring.

File: tests/test_show_version.py
Comment (kwinkunks) [reply tests/test_show_version.py:31]: Done

File: tests/conftest.py
Comment (kwinkunks) [reply tests/conftest.py:10]: Again, tidied up with more explicit language.

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:102]: This indentation is standard for Google-style docstrings (this package does not have a standard yet though). I actually assumed `black` would do whatever it wants but it seems not to touch docstrings?

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:41]: Reworded to deal with explanation of why this function exists.

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:38]: Lol, flake8 fails with W293 (blank line has whitespace) if you have a blank line in the middle of a string literal.

**Comment by kwinkunks** [comment]
Thanks for the reviews, all!

File: komodo/show_version.py
Comment (hnformentin) [reply komodo/show_version.py:38]: Thanks! There is still one extra \ by the end of the message.

File: komodo/show_version.py
Comment (hnformentin) [inline komodo/show_version.py:137]: Missing a blank space in `file{path_str}`, correct?

File: komodo/show_version.py
Comment (hnformentin) [reply komodo/show_version.py]: I understand. In this case, what do you think about adding this nomenclature in the [release manager manual](https://github.com/equinor/scout/blob/2e2086ec59fae3492f098b8114f1e83d9f79fcf2/release-process/README.md)? If you think it is a good idea, just creating the issue would be suffice since solving it is not exactly on the scope of this PR.

File: tests/conftest.py
Comment (hnformentin) [reply tests/conftest.py]: Your explanation here is really clear and indeed, this difference is not evident. 
I think they could be arranged in the docstrings of `mock_komodo_env_vars` and `mock_komodoenv_env_vars`, for example, and that would probably be useful for a future developer.

File: tests/test_show_version.py
Comment (hnformentin) [reply tests/test_show_version.py]: That I did not know, thanks for considering.

**Comment by kwinkunks** [comment]
Just realized I did not add the tool to the README. Done, with a small clarification to the CLI help string as well.

File: tests/conftest.py
Comment (kwinkunks) [reply tests/conftest.py]: Added some clarification to `conftest.py`.

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py]: done

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:38]: Right... I was trying to sidestep a `pylint` issue, but in a stupid way. Fixed.

File: komodo/show_version.py
Comment (kwinkunks) [reply komodo/show_version.py:137]: It does look weird but it's correct, because sometimes the sentence ends there.

**Review by hnformentin — APPROVED** [review]
Looks good, appreciate considering all the comments :-)

**Pull request merged**
